// @digitalcredentials/x25519-key-agreement-2020-context/dist/context.esm.js
function unwrapExports (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var context = createCommonjsModule(function (module, exports) {
/*! For license information please see context.js.LICENSE.txt */
(()=>{var e={717:e=>{e.exports={CONTEXT_FILENAME:"x25519-key-agreement-2020-v1.jsonld",CONTEXT_URL:"https://w3id.org/security/suites/x25519-2020/v1",CBORLD_VALUE:23};},869:e=>{e.exports={"@context":{id:"@id",type:"@type","@protected":!0,X25519KeyAgreementKey2020:{"@id":"https://w3id.org/security#X25519KeyAgreementKey2020","@context":{"@protected":!0,id:"@id",type:"@type",controller:{"@id":"https://w3id.org/security#controller","@type":"@id"},revoked:{"@id":"https://w3id.org/security#revoked","@type":"http://www.w3.org/2001/XMLSchema#dateTime"},publicKeyMultibase:{"@id":"https://w3id.org/security#publicKeyMultibase","@type":"https://w3id.org/security#multibase"}}}}};},858:(e,t,r)=>{const{CONTEXT_URL:o}=r(717),s=r(869);e.exports={documentLoader(e){if(e!==o)throw new Error(`Loading document "${e}" is not allowed.`);return {contextUrl:null,document:s,documentUrl:e}}};},243:(e,t,r)=>{const o=r(869),s=r(717),{documentLoader:i}=r(858),{CONTEXT_URL:d,CBORLD_VALUE:n}=s,p=new Map;p.set(s.CONTEXT_URL,o);const c=new Map;c.set(d,n),e.exports={constants:s,contexts:p,appContextMap:c,documentLoader:i,CONTEXT_URL:d,CONTEXT:o};}},t={},r=function r(o){var s=t[o];if(void 0!==s)return s.exports;var i=t[o]={exports:{}};return e[o](i,i.exports,r),i.exports}(243),o=exports;for(var s in r)o[s]=r[s];r.__esModule&&Object.defineProperty(o,"__esModule",{value:!0});})();
});

var context$1 = unwrapExports(context);
var context_1 = context.contexts;
var context_2 = context.constants;
var context_3 = context.CONTEXT;
var context_4 = context.CONTEXT_URL;
var context_5 = context.appContextMap;
var context_6 = context.documentLoader;

export default context$1;
export { context_3 as CONTEXT, context_4 as CONTEXT_URL, context_5 as appContextMap, context_2 as constants, context_1 as contexts, context_6 as documentLoader };
