// @digitalcredentials/sign-and-verify-core/dist/esm/verifier.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { getCustomLoader, addDidDocuments, getPreloadedAssertionMethods } from "./common";
import { Ed25519Signature2020 } from '@digitalcredentials/ed25519-signature-2020';
import { Ed25519VerificationKey2020 } from '@digitalcredentials/ed25519-verification-key-2020';
import { X25519KeyAgreementKey2020 } from '@digitalcredentials/x25519-key-agreement-key-2020';
import { CryptoLD } from 'crypto-ld';
import * as vc from '@digitalcredentials/vc';
import * as didWeb from '@interop/did-web-resolver';
import * as didKey from '@digitalcredentials/did-method-key';
// NOTE: This method is a simple and common issuer validation
// You may modify this method to suit the validation needs
// of your organization
export const validateCredential = (verifiableCredential, issuerMembershipRegistry) => __awaiter(void 0, void 0, void 0, function* () {
    if (typeof verifiableCredential.issuer === 'object') {
        return issuerMembershipRegistry.hasOwnProperty(verifiableCredential.issuer.id);
    }
    return issuerMembershipRegistry.hasOwnProperty(verifiableCredential.issuer);
});
export const validatePresentation = (verifiablePresentation, issuerMembershipRegistry) => __awaiter(void 0, void 0, void 0, function* () {
    if (!verifiablePresentation.verifiableCredential) {
        // presentation may omit credential
        return true;
    }
    if (Array.isArray(verifiablePresentation.verifiableCredential)) {
        return verifiablePresentation.verifiableCredential.every((credential) => {
            return validateCredential(credential, issuerMembershipRegistry);
        });
    }
    return validateCredential(verifiablePresentation.verifiableCredential, issuerMembershipRegistry);
});
export const createVerifier = (preloadedDidDocuments) => {
    const cryptoLd = new CryptoLD();
    cryptoLd.use(Ed25519VerificationKey2020);
    cryptoLd.use(X25519KeyAgreementKey2020);
    const didWebDriver = didWeb.driver({ cryptoLd });
    const didKeyDriver = didKey.driver();
    let customLoaderProto = getCustomLoader();
    customLoaderProto = addDidDocuments(customLoaderProto, preloadedDidDocuments);
    customLoaderProto.addResolver({
        ['did:web:']: {
            resolve: (_did) => __awaiter(void 0, void 0, void 0, function* () {
                return didWebDriver.get({ did: _did });
            }),
        },
        ['did:key:']: {
            resolve: (_did) => __awaiter(void 0, void 0, void 0, function* () {
                return didKeyDriver.get({ did: _did });
            }),
        },
    });
    let transmuteLoader = customLoaderProto.buildDocumentLoader();
    const preloadedAssertionMethods = getPreloadedAssertionMethods(preloadedDidDocuments);
    const customLoader = (url) => __awaiter(void 0, void 0, void 0, function* () {
        const result = preloadedAssertionMethods.get(url);
        if (result) {
            const document = Object.assign({ '@context': 'https://w3id.org/security/suites/ed25519-2020/v1' }, result);
            return {
                documentUrl: url,
                document
            };
        }
        return transmuteLoader(url);
    });
    function verify({ verifiableCredential, issuerMembershipRegistry, options }) {
        return __awaiter(this, void 0, void 0, function* () {
            // During verification, the public key is fetched via documentLoader,
            // so no key is necessary when creating the suite
            const suite = new Ed25519Signature2020();
            try {
                const result = yield vc.verifyCredential({
                    credential: verifiableCredential,
                    documentLoader: customLoader,
                    suite
                });
                const verified = result.verified;
                const valid = yield validateCredential(verifiableCredential, issuerMembershipRegistry);
                return Object.assign(Object.assign({}, result), { verified, valid });
            }
            catch (e) {
                console.error(e);
                throw e;
            }
        });
    }
    function verifyPresentation({ verifiablePresentation, issuerMembershipRegistry, options }) {
        return __awaiter(this, void 0, void 0, function* () {
            // During verification, the public key is fetched via documentLoader,
            // so no key is necessary when creating the suite
            const suite = new Ed25519Signature2020();
            const toVerify = {
                presentation: Object.assign({}, verifiablePresentation),
                documentLoader: customLoader,
                suite: suite
            };
            if (options && options.challenge) {
                toVerify['challenge'] = options.challenge;
            }
            const result = yield vc.verify(toVerify);
            const verified = result.verified;
            const valid = yield validatePresentation(verifiablePresentation, issuerMembershipRegistry);
            return Object.assign(Object.assign({}, result), { verified, valid });
        });
    }
    return {
        verify,
        verifyPresentation
    };
};
